scipy/stats/tests/test_qmc.py

```diff
@@ -0,0 +1,621 @@
+        sample = np.array([[2, 1, 1, 2, 2, 2],
+                           [1, 2, 2, 2, 2, 2],
+                           [2, 1, 1, 1, 1, 1],
+                           [1, 1, 1, 1, 2, 2],
+                           [1, 2, 2, 2, 1, 1],
+                           [2, 2, 2, 2, 1, 1],
+                           [2, 2, 2, 1, 2, 2]])
+        sample = (2.0 * sample - 1.0) / (2.0 * 2.0)
+
+        assert_allclose(qmc.discrepancy(sample, method='MD'), 2.5000,
+                        atol=1e-4)
+        assert_allclose(qmc.discrepancy(sample, method='WD'), 1.3680,
+                        atol=1e-4)
+        assert_allclose(qmc.discrepancy(sample, method='CD'), 0.3172,
+                        atol=1e-4)
+        assert_allclose(qmc.discrepancy(sample, method='star'), 0.037451,
+                        atol=1e-4)
+
+        assert_raises(ValueError, qmc.discrepancy, sample, False, 'toto')
+
+    def test_update_discrepancy(self):
+        space_1 = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+        space_1 = (2.0 * space_1 - 1.0) / (2.0 * 6.0)
+
+        disc_init = qmc.discrepancy(space_1[:-1], iterative=True)
+        disc_iter = _update_discrepancy(space_1[-1], space_1[:-1],
+                                        disc_init)
+
+        assert_allclose(disc_iter, 0.0081, atol=1e-4)
+
+    def test_perm_discrepancy(self):
+        doe_init = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+        doe_init = (2.0 * doe_init - 1.0) / (2.0 * 6.0)
+
+        disc_init = qmc.discrepancy(doe_init)
+
+        row_1, row_2, col = 5, 2, 1
+
+        doe = copy.deepcopy(doe_init)
+        doe[row_1, col], doe[row_2, col] = doe[row_2, col], doe[row_1, col]
+
+        disc_valid = qmc.discrepancy(doe)
+        disc_perm = _perturb_discrepancy(doe_init, row_1, row_2, col,
+                                         disc_init)
+
+        assert_allclose(disc_valid, disc_perm)
+
+    def test_n_primes(self):
+        primes = n_primes(10)
+        assert primes[-1] == 29
+
+        primes = n_primes(168)
+        assert primes[-1] == 997
+
+        primes = n_primes(350)
+        assert primes[-1] == 2357
+
+    def test_primes(self):
+        primes = primes_from_2_to(50)
+        out = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]
+        assert_allclose(primes, out)
+
+
+class TestQMC(object):
+    def test_van_der_corput(self):
+        sample = van_der_corput(10)
+        out = [0., 0.5, 0.25, 0.75, 0.125, 0.625, 0.375, 0.875, 0.0625, 0.5625]
+        assert_almost_equal(sample, out)
+
+        sample = van_der_corput(7, start_index=3)
+        assert_almost_equal(sample, out[3:])
+
+    def test_van_der_corput_scramble(self):
+        out = van_der_corput(10, scramble=True, seed=123456)
+
+        sample = van_der_corput(7, start_index=3, scramble=True,
+                                seed=123456)
+        assert_almost_equal(sample, out[3:])
+
+    def test_halton(self):
```

Comment:
For the tests that are similar among the classes, can they be combined? One way to do that is with `pytest.parametrize`. In this case, though, I'd suggest creating a common suite of tests in a class e.g. `QMCEngineTests` in terms of an attribute `self.qmce = None`, then subclasses of `QMCEngineTests` define `self.qmce` as one of the subclasses of `QMCEngine`, e.g. `class TestSobol(QMCEngine)` defines `self.qmce = Sobol`. Because of the name, `QMCEngineTests` doesn't actual run; only the subclasses actually run. Sorry if that's confusing. I can write it out in more than one sentence if you're interested : )

Reply:
Yup! Something like [this](https://github.com/scipy/scipy/blob/71c291649aec80e1fbe762c7a81ea2c5925bfeb9/scipy/optimize/tests/test_linprog.py#L266) is what I had in mind. You can see that tests are defined in terms of `self.method`, then that actually gets set in a [subclass](https://github.com/scipy/scipy/blob/71c291649aec80e1fbe762c7a81ea2c5925bfeb9/scipy/optimize/tests/test_linprog.py#L1533), and finally the tests run in a [subclass of that](https://github.com/scipy/scipy/blob/71c291649aec80e1fbe762c7a81ea2c5925bfeb9/scipy/optimize/tests/test_linprog.py#L1612).

Reply:
I was looking into this and I actually don't see lot of rewrite here. The tests are quite different from one method to the other. I improved the setup handling of the Sobol' tests, but here I don't really see what I could do better.

Reply:
My comment probably should have gone in `test_sobol_.py`, but I think that many of the tests that I had in mind there apply to the other `QMCEngine` subclasses, too. 

I will close this comment and open others in `test_sobol_.py`

Reply:
So here is my proposal based on what you said.

```python
class QMCEngineTests:
    qmce = ...
    can_scramble = ...
    sample_unscrambled_1d = ...
    sample_scrambled_1d = ...
    sample_unscrambled_3d = ...
    sample_scrambled_3d = ...

    scramble = [True, False]

    @pytest.mark.parametrize("scramble", scramble)
    def test_0dim(self, scramble):
        if self.can_scramble:
            engine = self.qmce(d=0, scramble=scramble)
        else:
            if scramble:
                engine = self.qmce(d=0)
            else:
                pytest.skip()
        sample = engine.random(4)
        assert_array_equal(np.empty((4, 0)), sample)

    @pytest.mark.parametrize("scramble", scramble)
    def test_bounds(self, scramble):
        if self.can_scramble:
            engine = self.qmce(d=100, scramble=scramble)
        else:
            if scramble:
                engine = self.qmce(d=100)
            else:
                pytest.skip()
        sample = engine.random(1024)
        assert_(np.all(sample >= 0))
        assert_(np.all(sample <= 1))


class TestHalton(QMCEngineTests):
    qmce = qmc.Halton
    can_scramble = True

class TestLHS(QMCEngineTests):
    qmce = qmc.LatinHypercube
    can_scramble = False
```

It works but I don't really like the code reuse around the scramble... Do you see a better way? I can start with that otherwise if you're ok with it.

Reply:
@rgommers I am pinging you in case you have an opinion about this.

Reply:
The combination `can_scramble=False, scramble=True` sounds like it should raise (from a UX perspective, not knowing what it does). If that comment doesn't make sense, please ignore.

The double `if-else` seems to be repeated in `test_0dim` and `test_bounds` so could go in a separate method (e.g., `get_engine(self, can_scramble, scramble)`).

Reply:
Ah right, using a simple method like `get_engine` would do it. Thanks! I was looking into more complicated things like parametric fixture and did not see it was so simple 😅.

`can_scramble` might not be the best name. It is to say that the underlying engine accept or not a parameter named `scramble`. But you're right, in the case where there is no scramble, the it makes more sense to only have the unscrambled case. It originally did this as the default for classes with scramble is True. I will change this, thanks again.

Reply:
@mdhaber I updated some tests using this proposal. Please let me know if it is ok for you.

Reply:
I'll take a look when I review the QMCEngines.